ns1/record_helper.go

```diff
@@ -0,0 +1,74 @@
+package ns1
+
+import (
+	"github.com/hashicorp/terraform/helper/hashcode"
+	"github.com/hashicorp/terraform/helper/mutexkv"
+	ns1 "gopkg.in/ns1/ns1-go.v2/rest"
+	"gopkg.in/ns1/ns1-go.v2/rest/model/dns"
+	"log"
+	"strconv"
+)
+
+type RecordMutexKV struct {
+	mutexkv *mutexkv.MutexKV
+	tracker map[string][]interface{}
+}
+
+func (m *RecordMutexKV) Lock(client *ns1.Client, i interface{}, record string, zone string) error {
+	log.Printf("[DEBUG] Locking Record %q", record)
+	hashRecord := strconv.Itoa(hashcode.String(record)) + "." + zone
+	m.mutexkv.Lock(hashRecord)
+	m.mutexkv.Lock(record)
+	defer m.mutexkv.Unlock(record)
```

Comment:
do we also need to unlock hashRecord?

Reply:
no. this is a bit confusing and it took me a very long time to page back in what this was all about. let me try to explain...

Reply:
there are three levels of locking going on when `Lock()` in `record_helper.go` gets called:
  1. a universal lock on the record which is implemented via the creation of a `TXT` record. this makes it so two simultaneous `terraform apply`'s on the same ns1 record won't step on top of each other.
  1. a mutex lock implemented via a `m.mutexkv.Lock(hashRecord)` in the `Lock()` function which makes it so the same `terraform apply` that would affect the same record (i.e. applying changes to two `ns1_answer` resources on the same record) won't step on top of each other.
  1. a smaller mutex lock, which is more of a meta-lock, on the record that is implemented via the `m.mutexkv.Lock(record)` operation. it is a mutex on `tracker` since only one thing ought to be modifying `tracker` at once. the `tracker` map simply keeps track of how many resources are making changes to the record, so when `Unlock()` is called it knows when to delete the `TXT` record (when there are 0 things in the `tracker` map for that record.

so with all that said, `defer m.mutexkv.Unlock(record)` here is unlocking the mutex on `record`. the mutex lock on `hashRecord` must remain in place even after `Lock()` returns and should only be unlocked when `Unlock()` is called.